ABCI++ spec. First complete version (#366)
tendermint/spec: spec/abci++/abci++_app_requirements_002_draft.md

---
order: 3
title: Application Requirements
---

# Application Requirements

This section specifies what Tendermint expects from the Application. It is structured as a set
of formal requirement that can be used for testing and verification of the Application's logic.

Let $p$ and $q$ be two different correct proposers in rounds $r_p$ and $r_q$ respectively, in height $h$.
Let $s_{p,h-1}$ be $p$'s Application's state committed for height $h-1$.
Let $v_p$ (resp. $v_q$) be the block that $p$'s (resp. $q$'s) Tendermint passes on to the Application
via `RequestPrepareProposal` as proposer of round $r_p$ (resp $r_q$), height $h$, also known as the
raw proposal.
Let $v'_p$ (resp. $v'_q$) the possibly modified block $p$'s (resp. $q$'s) Application returns via
`ResponsePrepareProposal` to Tendermint, also known as the prepared proposal.

Process $p$'s prepared proposal can differ in two different rounds where $p$ is the proposer.

* Requirement 1 [`PrepareProposal`, header-changes] When the blockchain is in same-block execution mode,
  $p$'s Application provides values for the following parameters in `ResponsePrepareProposal`:

>   _AppHash_, _TxResults_, _ConsensusParams_, _ValidatorUpdates_.

Reviewer: > I see. I believe that by currently you mean the params in EndBlock. Please let me know if my assumption isn't right.

correct!

raw proposal.
Let $v'_p$ (resp. $v'_q$) the possibly modified block $p$'s (resp. $q$'s) Application returns via
`ResponsePrepareProposal` to Tendermint, also known as the prepared proposal.

Process $p$'s prepared proposal can differ in two different rounds where $p$ is the proposer.

* Requirement 1 [`PrepareProposal`, header-changes] When the blockchain is in same-block execution mode,
  $p$'s Application provides values for the following parameters in `ResponsePrepareProposal`:
  _AppHash_, _TxResults_, _ConsensusParams_, _ValidatorUpdates_. Provided values for
  _ConsensusParams_ and _ValidatorUpdates_ MAY be empty to denote that the Application
  wishes to keep the current values.

Parameters _AppHash_, _TxResults_, _ConsensusParams_, and _ValidatorUpdates_ are used by Tendermint to
compute various hashes in the block header that will finally be part of the proposal.

* Requirement 2 [`PrepareProposal`, no-header-changes] When the blockchain is in next-block execution
  mode, $p$'s Application does not provide values for the following parameters in `ResponsePrepareProposal`:
  _AppHash_, _TxResults_, _ConsensusParams_, _ValidatorUpdates_.

> In practical terms, Requirement 2 implies that Tendermint will ignore those parameters in `ResponsePrepareProposal`.

Author: We can certainly log an error on Tendermint side if an App operating in next-block execution mode is filling these params; however, by the nature of ABCI (and ABCI++) we cannot report it to the App.
Moreover, an App filling up these fields, while in next-block execution mode, shows it's not ready for being released...

> In practical terms, Requirement 2 implies that Tendermint will ignore those parameters in `ResponsePrepareProposal`.

Reviewer 2: My point is this should be considered an error, not just ignore it. If, for example, the error handling policy for ABCI++ is to just panic tendermint in such cases, then it should panic.

  mode, $p$'s Application does not provide values for the following parameters in `ResponsePrepareProposal`:
  _AppHash_, _TxResults_, _ConsensusParams_, _ValidatorUpdates_.

In practical terms, Requirements 1 and 2 imply that Tendermint will (a) panic if the Application is in
same-block execution mode and _does not_ provide values for
_AppHash_, _TxResults_, _ConsensusParams_, and _ValidatorUpdates_, or
(b) log an error if the Application is in next-block execution mode and _does_ provide values for
_AppHash_, _TxResults_, _ConsensusParams_, or _ValidatorUpdates_ (the values provided will be ignored).

* Requirement 3 [`PrepareProposal`, timeliness] If $p$'s Application fully executes prepared blocks in
  `PrepareProposal` and the network is in a synchronous period while processes $p$ and $q$ are in $r_p$,

>   then the value of _TimeoutPropose_ at $q$ must be such that $q$ does not prevote _nil_ in $r_p$ due to
>   $q$'s propose timer timing out.

Reviewer 3: ```suggestion
  then the value of _TimeoutPropose_ at $q$ must be such that $q$'s propose timer does not time out (which would result in $q$ prevoting _nil_ in $r_p$).
  .
```

>**TODO** I have left out all the "events" as they don't have any impact in safety or liveness
>(same for consensus params, and validator set)

* Requirement 12 [`FinalizeBlock`, determinism-1]: For any correct process $p$,
  $s_{p,h}$ exclusively depends on $s_{p,h-1}$ and $v_{p,h}$.

* Requirement 13 [`FinalizeBlock`, determinism-2]: For any correct process $p$,
  the contents of $T_{p,h}$ exclusively depend on $s_{p,h-1}$ and $v_{p,h}$.

> Note that Requirements 12 and 13, combined with Agreement property of consensus ensure

Reviewer 3: I guess this renumbering was a pain. Also for future extensions, perhaps we should give names to the requirements such that adding new ones does not change the names of the old ones.

Author: Good idea. Indeed, now that we are merging, I don't think it'll be a good idea to re-number again, so what you propose is probably what will happen if we need to extend the properties after this PR is merged.
